File: src/ros/master.cpp
#include "torch-ros.h"
#include <ros/master.h>

ROSIMP(bool, Master, execute)(
  const char *method,
  xamla::Variable *request,
  xamla::Variable *response,
  xamla::Variable *payload,
  bool wait_for_master) {

  XmlRpc::XmlRpcValue request_;
  VariableToXmlRpcValue(*request, request_);    // copy request

  XmlRpc::XmlRpcValue response_, payload_;
  bool result = ros::master::execute(method, request_, response_, payload_, wait_for_master);

  XmlRpcValueToVariable(response_, *response);  // copy response
  XmlRpcValueToVariable(payload_, *payload);
  return result;
}

ROSIMP(const char*, Master, getHost)() {
  return ros::master::getHost().c_str();
}

ROSIMP(int, Master, getPort)() {
  return static_cast<int>(ros::master::getPort());
}

ROSIMP(const char*, Master, getURI)() {
  return ros::master::getURI().c_str();
}

ROSIMP(bool, Master, check)() {
  return ros::master::check();
}

ROSIMP(bool, Master, getTopics)(xamla::VariableTable_ptr *output) {
  if (!*output) {
    output->reset(new xamla::VariableTable());
  }

  xamla::VariableTable& output_table = **output;

  // call to maste api
  ros::master::V_TopicInfo topics;
  bool result = ros::master::getTopics(topics);

  // store result in table
  ros::master::V_TopicInfo::const_iterator i = topics.begin();
  for (; i != topics.end(); ++i) {
    const ros::master::TopicInfo& topic = *i;
    xamla::VariableTable_ptr t(new xamla::VariableTable());
    (*t)["topic"] = topic.name;
    (*t)["datatype"] = topic.datatype;
    output_table[topic.name] = t;
  }
  return result;
}

ROSIMP(bool, Master, getNodes)(std::vector<std::string> *output) {
  return ros::master::getNodes(*output);
}

ROSIMP(void, Master, setRetryTimeout)(int sec, int nsec) {
  ros::master::setRetryTimeout(ros::WallDuration(sec, nsec));
}

File: master.lua
local ffi = require 'ffi'
local torch = require 'torch'
local ros = require 'ros.env'
local utils = require 'ros.utils'
local std = ros.std

local master = {}
ros.master = master

function init()
  local names = {
    'execute',
    'getHost',
    'getPort',
    'getURI',
    'check',
    'getTopics',
    'getNodes',
    'setRetryTimeout'
  }

  return utils.create_method_table("ros_Master_", names)
end

local f = init()

function master.execute(method, request, wait_for_master)
  if not torch.isTypeOf(request, std.Variable) then
    request = std.Variable(request)
  end
  local response, payload = std.Variable(), std.Variable()
  local result = f.execute(method, request:cdata(), response:cdata(), payload:cdata(), wait_for_master or false)
  return result, response:get(), payload:get()
end

function master.getHost()
  return ffi.string(f.getHost())
end

function master.getPort()
  return f.getPort()
end

function master.getURI()
  return ffi.string(f.getURI())
end

function master.check()
  return f.check()
end

function master.getTopics(output)
  local v = output or std.VariableTable()
  f.getTopics(v:cdata())
  return v:totable()
end

function master.getNodes(output)
  local v = output or std.StringVector()
  f.getNodes(v:cdata())
  return v
end

function master.setRetryTimeout(_1, _2)
  local d = _1
  if not torch.isTypeOf(d, ros.Duration) then
    d = ros.Duration(_1, _2)
  end
  f.setRetryTimeout(d:get_sec(), d:get_nsec())
end
